src/html5/parser/document.rs

```diff
@@ -618,6 +603,54 @@ impl DocumentHandle {
     pub fn has_cyclic_reference(&self, node_id: NodeId, parent_id: NodeId) -> bool {
         self.get().has_cyclic_reference(node_id, parent_id)
     }
+
+    fn insert_id_attribute(&mut self, value: &str, element_id: NodeId) -> Result<()> {
+        if !is_valid_id_attribute_value(value) {
+            return Err(Error::DocumentTask(format!(
+                "Attribute value '{}' did not pass validation",
+                value
+            )));
+        }
+
+        // an ID must be tied to only one element
+        if self.get().named_id_elements.contains_key(value) {
+            return Err(Error::DocumentTask(format!(
+                "ID '{}' already exists in DOM",
+                value
+            )));
+        }
+
+        let mut doc = self.get_mut();
+        let data = &mut doc
+            .get_node_by_id_mut(element_id)
+            .ok_or(Error::DocumentTask(format!(
+                "Node ID {} not found",
+                element_id
+            )))?
+            .data;
+
+        if let NodeData::Element(_) = data {
+        } else {
+            return Err(Error::DocumentTask(format!(
+                "Node ID {} is not an element",
+                element_id
+            )));
+        }
```

Comment:
In a later PR, we can consolidate lines 634-637 with line 646, so that there's a single `if let` statement.

Reply:
This is what I was thinking but wasn't sure if a return statement was allowed there since we were assigning to a variable (I didn't attempt it, though.) If that's possible then I can switch it

Reply:
I just checked, and it looks like Rust will allow it.